File: src/islands.c
// ISLANDS
// Attila Bagyoni, 2018
// This software is public domain. Do whatever you want with it.

#include "lib/main.h"
#include "lib/perlin.h"

#define GRID_MARGINS 10
#define GRID_SIZE_IN_PX ((((WID > HEI) ? HEI : WID) - 2 * GRID_MARGINS) / 6)

#define THRESHOLDVAL1 195
#define THRESHOLDVAL2 190

unsigned char const background_base_color[] = {198, 151, 63};
unsigned char const border_color[] = {154, 115, 82};
unsigned char const foreground_base_color[] = {202, 168, 131};

unsigned char compute_threshold(unsigned char value, unsigned char replacement_1, unsigned char replacement_2,
                                unsigned char replacement_3) {
    if (value > THRESHOLDVAL1) {
        return replacement_1;
    } else {
        if (value > THRESHOLDVAL2) {
            return replacement_2;
        } else {
            return replacement_3;
        }
    }
}

void draw_horiz_dashed(int y) {
    int x;
    for (x = 0; x < WID; x++) {
        if (x % 20 < 10) {
            bytes[(WID * y + x) * 3] = 100;
            bytes[(WID * y + x) * 3 + 1] = 96;
            bytes[(WID * y + x) * 3 + 2] = 82;
        }
    }
}

void draw_vert_dashed(int x) {
    int y;
    for (y = 0; y < HEI; y++) {
        if (y % 20 < 10) {
            bytes[(WID * y + x) * 3] = 100;
            bytes[(WID * y + x) * 3 + 1] = 96;
            bytes[(WID * y + x) * 3 + 2] = 82;
        }
    }
}

void draw() {
    generate_noise();
    int x, y;
    for (x = 0; x < WID; x++) {
        for (y = 0; y < HEI; y++) {
            unsigned char val = 185 + (char) (fractal(x, y, 0.004, 8) * 70);
            bytes[(WID * y + x) * 3] = compute_threshold(val, foreground_base_color[0] * val / 255, border_color[0],
                                                         (background_base_color[0] + val) / 2);
            bytes[(WID * y + x) * 3 + 1] = compute_threshold(val, foreground_base_color[1] * val / 255, border_color[1],
                                                             (background_base_color[1] + val) / 2);
            bytes[(WID * y + x) * 3 + 2] = compute_threshold(val, foreground_base_color[2] * val / 255, border_color[2],
                                                             (background_base_color[2] + val) / 2);
        }
    }
    for (int y = GRID_MARGINS; y < HEI; y += GRID_SIZE_IN_PX) {
        draw_horiz_dashed(y);
    }
    for (int x = GRID_MARGINS; x < WID; x += GRID_SIZE_IN_PX) {
        draw_vert_dashed(x);
    }
}

File: wallpapers/src/algorithms/islands.rs
use crate::algorithms::Algorithm;
use crate::utils::perlin::Perlin;
use image::{Rgb, RgbImage};
use rand::Rng;

/// Islands
/// Based on the ISLANDS algorithm implementation by Attila Bagyoni, 2018
/// https://github.com/bagyoni/procedural-wallpapers
pub struct Islands {
    threshold_val_1: u8,
    threshold_val_2: u8,
    background_base_color: [u8; 3],
    border_color: [u8; 3],
    foreground_base_color: [u8; 3],
    dashed_grid_color: [u8; 3],
    grid_margins: usize,
}

impl Default for Islands {
    fn default() -> Self {
        Islands {
            threshold_val_1: 195,
            threshold_val_2: 190,
            background_base_color: [198, 151, 63],
            border_color: [154, 115, 82],
            foreground_base_color: [202, 168, 131],
            dashed_grid_color: [100, 96, 82],
            grid_margins: 10,
        }
    }
}

impl Islands {
    fn compute_threshold(
        &self,
        value: u8,
        replacement_1: u8,
        replacement_2: u8,
        replacement_3: u8,
    ) -> u8 {
        if value > self.threshold_val_1 {
            replacement_1
        } else if value > self.threshold_val_2 {
            replacement_2
        } else {
            replacement_3
        }
    }
    fn draw_horiz_dashed(&self, img: &mut RgbImage, y: usize) {
        for x in 0..img.width() {
            if x % 20 < 10 {
                let pixel = img.get_pixel_mut(x, y as u32);
                *pixel = Rgb(self.dashed_grid_color);
            }
        }
    }
    fn draw_vert_dashed(&self, img: &mut RgbImage, x: usize) {
        for y in 0..img.height() {
            if y % 20 < 10 {
                let pixel = img.get_pixel_mut(x as u32, y);
                *pixel = Rgb(self.dashed_grid_color);
            }
        }
    }
}

impl<R: Rng> Algorithm<R> for Islands {
    fn build(&mut self, rng: &mut R, img: &mut RgbImage) -> Result<(), String> {
        // Six grid squares for the longer dimension of the image
        let grid_size_in_px: usize = ((if img.width() > img.height() {
            img.height()
        } else {
            img.width()
        } as f64
            - (2.0 * self.grid_margins as f64))
            / 6.0) as usize;
        let mut perlin = Perlin::new(img.width() as usize, img.height() as usize);
        perlin.regenerate_noise(rng);
        for (x, y, pixel) in img.enumerate_pixels_mut() {
            let val: u8 = (185.0 + perlin.fractal(x as f64, y as f64, 0.004, 8)? * 70.0) as u8;
            *pixel = Rgb([
                self.compute_threshold(
                    val,
                    (self.foreground_base_color[0] as f64 * val as f64 / 255.0) as u8,
                    self.border_color[0],
                    ((self.background_base_color[0] as u32 + val as u32) as f64 / 2.0) as u8,
                ),
                self.compute_threshold(
                    val,
                    (self.foreground_base_color[1] as f64 * val as f64 / 255.0) as u8,
                    self.border_color[1],
                    ((self.background_base_color[1] as u32 + val as u32) as f64 / 2.0) as u8,
                ),
                self.compute_threshold(
                    val,
                    (self.foreground_base_color[2] as f64 * val as f64 / 255.0) as u8,
                    self.border_color[2],
                    ((self.background_base_color[2] as u32 + val as u32) as f64 / 2.0) as u8,
                ),
            ]);
        }
        for y in (self.grid_margins..(img.height() as usize)).step_by(grid_size_in_px) {
            self.draw_horiz_dashed(img, y);
        }
        for x in (self.grid_margins..(img.width() as usize)).step_by(grid_size_in_px) {
            self.draw_vert_dashed(img, x);
        }
        Ok(())
    }
}
